policy_agent/policy_agent.py

```diff
@@ -0,0 +1,655 @@
+        Remove the default reject all rule programmed by the plugin.
+        """
+        default_profile = DEFAULT_PROFILE_REJECT
+        try:
+            _log.info("Removing Default Profile")
+            _datastore_client.remove_profiles_from_endpoint(profile_names=[default_profile],
+                                                            endpoint_id=self.ep_id)
+            return True
+        except ProfileNotInEndpoint:
+            _log.info("Default Profile not on Endpoint")
+            return True
+        else:
+            _log.info("Default Profile Removal Failed")
+            return False
+
+
+class Endpoints(Resource):
+
+    def init_from_json(self, json):
+        self.kind = KIND_ENDPOINTS
+        self.name = json["metadata"]["name"]
+        self.namespace = json["metadata"]["namespace"]
+        self.subsets = json["subsets"]
+
+    @property
+    def key(self):
+        return "%s/%s" % (self.namespace, self.name)
+
+    def generate_svc_profiles_pods(self):
+        """
+        Endpoints objects contain a list of subsets
+        containing pod and policy info
+        subsets: [
+            {
+                addresses: [
+                    ...pods...
+                ],
+                ports: [
+                    ...associated ports...
+                ]
+            },
+            {
+                addresses: [
+                    ...pods...
+                ],
+                ports: [
+                    ...associated ports...
+                ]
+            },
+        ]
+        :return: A generated dict of profile-pod associations
+        :rtype: a dict of profiles mapping to a list of assoc pods
+        """
+        def verify_args(ports_obj, try_arg):
+            try:
+                arg = ports_obj[try_arg]
+            except KeyError:
+                arg = None
+            return arg
+
+        self.association_list = {}
+        for subset in self.subsets:
+            # Base profile_name string
+            profile_name = "%s_svc_%s" % (self.namespace, self.name)
+
+            # Generate a list of new svc profiles per port spec
+            port_rules = []
+            for port_spec in subset["ports"]:
+                dst_port = verify_args(port_spec, "port")
+                protocol = verify_args(port_spec, "protocol").lower()
+                if protocol and dst_port:
+                    port_rules.append(Rule(action="allow",
+                                           dst_ports=[dst_port],
+                                           protocol=protocol))
+                    profile_name = "%s_%s%s" % (
+                        profile_name, protocol, dst_port)
+
+            # Create Profile for the subset
+            if not _datastore_client.profile_exists(profile_name):
+                _log.info("Creating Profile %s" % profile_name)
+                profile_path = PROFILE_PATH % {"profile_id": profile_name}
+                _datastore_client.etcd_client.write(
```

Comment:
the write tags line is essentially what the datastore client uses to create a profile. I didn't want to use create_profile because it creates a default set of rules... however I did find a more efficient pycalico function to use so Ill throw it in